drivers/lora/sx1276.c

```diff
@@ -0,0 +1,494 @@
+	gpio_pin_write(dev_data.reset, GPIO_RESET_PIN, 0);
+
+	k_sleep(1);
+
+	gpio_pin_configure(dev_data.reset, GPIO_RESET_PIN,
+			   GPIO_DIR_IN | GPIO_PUD_NORMAL);
+	k_sleep(6);
+}
+
+void BoardCriticalSectionBegin(uint32_t *mask)
+{
+	*mask = irq_lock();
+}
+
+void BoardCriticalSectionEnd(uint32_t *mask)
+{
+	irq_unlock(*mask);
+}
+
+uint32_t RtcGetTimerElapsedTime(void)
+{
+	return counter_read(dev_data.counter);
+}
+
+u32_t RtcGetMinimumTimeout(void)
+{
+	/* TODO: Get this value from counter driver */
+	return 3;
+}
+
+void RtcSetAlarm(uint32_t timeout)
+{
+	struct counter_alarm_cfg alarm_cfg;
+
+	alarm_cfg.flags = 0;
+	alarm_cfg.ticks = timeout;
+
+	counter_set_channel_alarm(dev_data.counter, 0, &alarm_cfg);
+}
+
+uint32_t RtcSetTimerContext(void)
+{
+	return 0;
+}
+
+uint32_t RtcMs2Tick(uint32_t milliseconds)
+{
+	return counter_us_to_ticks(dev_data.counter, (milliseconds / 1000));
+}
+
+void DelayMsMcu(uint32_t ms)
+{
+	k_sleep(ms);
+}
+
+static void sx1276_irq_callback(struct device *dev,
+				struct gpio_callback *cb, u32_t pins)
+{
+	unsigned int i, pin;
+
+	pin = find_lsb_set(pins) - 1;
+
+	for (i = 0; i < SX1276_MAX_DIO; i++) {
+		if (pin == sx1276_dio_pins[i]) {
+			(*DioIrq[i])(NULL);
+		}
+	}
+}
+
+void SX1276IoIrqInit(DioIrqHandler **irqHandlers)
+{
+	unsigned int i;
+	static struct gpio_callback callbacks[SX1276_MAX_DIO];
+
+	/* Setup DIO gpios */
+	for (i = 0; i < SX1276_MAX_DIO; i++) {
+		dev_data.dio_dev[i] = device_get_binding(sx1276_dio_ports[i]);
+		if (dev_data.dio_dev[i] == NULL) {
+			LOG_ERR("Cannot get pointer to %s device",
+				sx1276_dio_ports[i]);
+			return;
+		}
+
+		gpio_pin_configure(dev_data.dio_dev[i], sx1276_dio_pins[i],
+				   GPIO_DIR_IN | GPIO_INT | GPIO_INT_EDGE |
+				   GPIO_INT_DEBOUNCE | GPIO_INT_ACTIVE_HIGH);
+
+		gpio_init_callback(&callbacks[i],
+				   sx1276_irq_callback,
+				   BIT(sx1276_dio_pins[i]));
+
+		if (gpio_add_callback(dev_data.dio_dev[i], &callbacks[i]) < 0) {
+			LOG_ERR("Could not set gpio callback.");
+			return;
+		}
+		gpio_pin_enable_callback(dev_data.dio_dev[i],
+					 sx1276_dio_pins[i]);
+	}
+
+}
+
+static int sx1276_transceive(u8_t reg, bool write, void *data, size_t length)
+{
+	const struct spi_buf buf[2] = {
+		{
+			.buf = &reg,
+			.len = 1
+		},
+		{
+			.buf = data,
+			.len = length
+		}
+	};
+	struct spi_buf_set tx = {
+		.buffers = buf,
+		.count = 2,
+	};
+
+	if (!write) {
+		const struct spi_buf_set rx = {
+			.buffers = buf,
+			.count = 2
+		};
+
+		return spi_transceive(dev_data.spi, &dev_data.spi_cfg,
+				&tx, &rx);
+	}
+
+	return spi_write(dev_data.spi, &dev_data.spi_cfg, &tx);
+}
+
+int sx1276_read(u8_t reg_addr, u8_t *data, u8_t len)
+{
+	return sx1276_transceive(reg_addr, false, data, len);
+}
+
+int sx1276_write(u8_t reg_addr, u8_t byte)
+{
+	return sx1276_transceive(reg_addr | BIT(7), true, &byte, 1);
+}
+
+void SX1276WriteBuffer(u16_t addr, u8_t *buffer, u8_t size)
+{
+	int ret, i;
+
+	for (i = 0; i < size; i++) {
```

Comment:
Please use the bulk transfer here. Same as the `SX1276ReadBuffer`

Reply:
Done, thanks